File: src/main/java/dev/akif/cookbook/recipe/RecipeInstructionRepository.java
package dev.akif.cookbook.recipe;

import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.ListCrudRepository;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface RecipeInstructionRepository extends ListCrudRepository<RecipeInstructionEntity, RecipeInstructionEntity.RecipeInstructionId> {
    @Query("""
           SELECT ri
           FROM recipe_instruction ri
           JOIN recipe r ON ri.recipeId = r.id
           WHERE r.id = :recipeId
           ORDER BY ri.sortOrder, ri.instruction
           """)
    List<RecipeInstructionEntity> findByRecipeId(Long recipeId);
}

File: src/main/kotlin/dev/akif/cookbook/recipe/instruction/RecipeInstructionRepository.kt
package dev.akif.cookbook.recipe.instruction

import kotlinx.coroutines.flow.Flow
import org.springframework.data.r2dbc.repository.Query
import org.springframework.data.repository.kotlin.CoroutineCrudRepository
import org.springframework.stereotype.Repository

@Repository
interface RecipeInstructionRepository: CoroutineCrudRepository<RecipeInstructionEntity, Long> {
    @Query(
        """SELECT ri.id, ri.recipe_id, ri.sort_order, ri.instruction
           FROM recipe_instructions ri JOIN recipes r ON ri.recipe_id = r.id
           WHERE r.id = :recipeId
           ORDER BY ri.sort_order, ri.instruction"""
    )
    fun findByRecipeId(recipeId: Long): Flow<RecipeInstructionEntity>
}
